openmls/src/group/managed_group/validation.rs

```diff
@@ -126,47 +108,78 @@ impl ManagedGroup {
-            Some(signature_public_key) => message
-                .plaintext
-                .verify_with_key(backend, &signature_public_key)?,
-            None => message.plaintext.verify(backend, &message.credential)?,
-        };
-
-        Ok(match plaintext.content() {
-            MlsPlaintextContentType::Application(application_message) => {
-                ProcessedMessage::ApplicationMessage(application_message.as_slice().to_vec())
+        // Add the context to the message and verify the membership tag if necessary
+        let serialized_context = self.group.context().tls_serialize_detached()?;
+
+        let context_plaintext = UnverifiedContextMessage::from_unverified_message_with_context(
+            unverified_message,
+            serialized_context,
+            self.group.epoch_secrets().membership_key(),
+            backend,
+        )?;
+
+        match context_plaintext {
+            UnverifiedContextMessage::Member(member_message) => {
+                // Signature verification
+                let verified_member_message =
+                    member_message.into_verified(backend, signature_key)?;
+
+                Ok(match verified_member_message.plaintext().content() {
+                    MlsPlaintextContentType::Application(application_message) => {
+                        ProcessedMessage::ApplicationMessage(
+                            application_message.as_slice().to_vec(),
+                        )
+                    }
+                    MlsPlaintextContentType::Proposal(_proposal) => {
+                        ProcessedMessage::ProposalMessage(Box::new(
+                            StagedProposal::from_mls_plaintext(
+                                self.ciphersuite(),
+                                backend,
+                                verified_member_message.take_plaintext(),
+                            )
+                            .unwrap(),
+                        ))
+                    }
+                    MlsPlaintextContentType::Commit(_commit) => {
+                        let staged_commit = self.group.stage_commit(
+                            verified_member_message.plaintext(),
+                            &self.proposal_store,
+                            &self.own_kpbs,
+                            None,
+                            backend,
+                        )?;
+                        ProcessedMessage::StagedCommitMessage(Box::new(staged_commit))
+                    }
+                })
             }
-            MlsPlaintextContentType::Proposal(_proposal) => {
-                ProcessedMessage::ProposalMessage(Box::new(
-                    StagedProposal::from_mls_plaintext(self.ciphersuite(), backend, plaintext)
-                        .unwrap(),
-                ))
+            UnverifiedContextMessage::External(external_message) => {
+                // Signature verification
+                match signature_key {
```

Comment:
Nitpicking again: Even though it makes returning the error slightly less explicit: I think I'd use an `ok_or` here. Or maybe an `if let`.